### DISCUSSION TITLE
Introduced unconditional hard dependency on `typing_extensions` (Starlette 0.35)

### DISCUSSION THREAD

**Discussion by palankai** [encode/starlette#2407] [Potential Issue] [answered]
Hi,

Starlette 0.35 cannot be used with Python> 3.9 because of hard dependency on not installed typed_extensions.

The `typing_extension` package is a conditional requirement according to `pyproject.toml`
https://github.com/encode/starlette/blob/0.35.0/pyproject.toml#L32

However, application.py requires it unconditionally:
https://github.com/encode/starlette/blob/0.35.0/starlette/applications.py#L6

I suggest to follow how it is imported in authentication.py:
https://github.com/encode/starlette/blob/0.35.0/starlette/authentication.py#L7

The error:
```
Traceback (most recent call last):
  File "/opt/src/main.py", line 7, in <module>
    from starlette.applications import Starlette
  File "/usr/local/lib/python3.12/site-packages/starlette/applications.py", line 6, in <module>
    from typing_extensions import ParamSpec
ModuleNotFoundError: No module named 'typing_extensions'
```

The bug (I think that can be classified as such) was introduced:
https://github.com/encode/starlette/commit/866a15f79c7f15b5dba695ed2180922b7191a19e

My application is a fairly "vanilla" small Startlette-based API running in docker with Python 3.12.1

A possible workaround is installing `typing_extensions` separately.
(I'd rather pin down Startlett for the time being `starlette!=0.35.0` for project using Python >= 3.10)

**Comment by Kludex** [comment]
PR welcome.

**Reply by Kludex** [reply, marked as answer]
Released: https://github.com/encode/starlette/pull/2410

Sorry for the trouble.
